+++ /usr/lib/python2.6/site-packages/nova/virt/libvirt/driver.py
@@ -2972,13 +2972,14 @@
         # at the same time. We're not trying to second guess which
         # those versions are. We'll just let libvirt report the
         # errors appropriately if the user enables both.
-
+        add_video_driver = False
         if CONF.vnc_enabled and CONF.libvirt_type not in ('lxc', 'uml'):
             graphics = vconfig.LibvirtConfigGuestGraphics()
             graphics.type = "vnc"
             graphics.keymap = CONF.vnc_keymap
             graphics.listen = CONF.vncserver_listen
             guest.add_device(graphics)
+            add_video_driver = True
 
         if CONF.spice.enabled and \
                 CONF.libvirt_type not in ('lxc', 'uml', 'xen'):
@@ -2987,6 +2988,23 @@
             graphics.keymap = CONF.spice.keymap
             graphics.listen = CONF.spice.server_listen
             guest.add_device(graphics)
+            add_video_driver = True
+
+        if add_video_driver:
+            VALID_VIDEO_DEVICES = ("vga", "cirrus", "vmvga", "xen", "qxl")
+            video = vconfig.LibvirtConfigGuestVideo()
+            meta_prop = image_meta.get('properties', {}) if image_meta else {}
+
+            if meta_prop.get('hw_video_model'):
+                video.type = meta_prop.get('hw_video_model')
+                if (video.type not in VALID_VIDEO_DEVICES):
+                    raise exception.InvalidVideoMode(model=video.type)
+
+            elif CONF.spice.enabled:
+                video.type = 'qxl'
+            if guest.os_type == vm_mode.XEN:
+                video.type = 'xen'
+            guest.add_device(video)
 
         # Qemu guest agent only support 'qemu' and 'kvm' hypervisor
         if CONF.libvirt_type in ('qemu', 'kvm'):